lib/rafka.rb

```diff
@@ -14,11 +14,35 @@ module Rafka
     port: 6380
   }.freeze
 
+  # Server errors upon which we should retry the operation.
+  RETRIABLE_ERRORS = {
+    "CONS Server shutdown" => true,
+  }.freeze
+
+  # Wraps errors from redis-rb to our own error classes
   def self.wrap_errors
     yield
   rescue Redis::CommandError => e
     raise ProduceError, e.message[5..-1] if e.message.start_with?("PROD ")
     raise ConsumeError, e.message[5..-1] if e.message.start_with?("CONS ")
     raise CommandError, e.message
   end
+
+  # Retries the operation upon Redis::CommandError
+  def self.with_retry(max_retries=5)
+    retries = 0
+
+    begin
+      yield
+    rescue Redis::CommandError => e
+      if RETRIABLE_ERRORS[e.message] && retries < max_retries
+        sleep 1
+        retries += 1
+        puts "retrying... #{retries} #{max_retries}"
+        retry
```

Comment:
I've considered an exponential backoff approach but concluded that with rafka's shutdown process in mind, a fixed interval is sufficient. That is because this is not a case of service congestion (e.g. rafka server is overloaded so we want to backoff) but it's a process that we know takes a fixed amount of time (1-5sec.)

Reply:
Regarding the "configurable max_retries" suggestion, I see that it _might_ be useful, but since it's not so straightforward how we should expose it (a separate option vs. reading from `:redis` config that already has a retry option), I say we do it when and if it's actually needed.

Reply:
I understand your rationale regarding the exponential back-off approach. Let's move on a fixed interval for now as you suggested.

However, I'd argue that the retry settings are not actually needed to be configurable. Rafka is already used in various types of services, with different requirements and workload. So, having a fixed retry interval for all rafka-rb clients does not scale. On the contrary it would be preferable to be able to set the back-off interval per service, based on the its workload and requirements.

Reply:
Agreed to do this in a second iteration and go with a default of 15secs. for now.